subprojects/build-plugins/src/main/kotlin/org/gradle/kotlin/dsl/build/tasks/GenerateKotlinDslApiExtensions.kt

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.build.tasks
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.CacheableTask
+import org.gradle.api.tasks.Classpath
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.InputFiles
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.api.tasks.PathSensitive
+import org.gradle.api.tasks.PathSensitivity
+import org.gradle.api.tasks.TaskAction
+
+import org.gradle.internal.classloader.ClassLoaderFactory
+import org.gradle.internal.classpath.DefaultClassPath
+
+import org.gradle.kotlin.dsl.*
+
+import org.gradle.kotlin.dsl.codegen.generateKotlinDslApiExtensionsSourceTo
+import org.gradle.kotlin.dsl.support.serviceOf
+
+import java.io.Closeable
+import java.io.File
+import java.util.*
+
+
+/**
+ * Generates Kotlin Extensions sources for API the Gradle Kotlin DSL.
+ */
+@CacheableTask
+@Suppress("MemberVisibilityCanBePrivate")
+open class GenerateKotlinDslApiExtensions : DefaultTask() {
+
+    init {
+        description = "Generates Gradle Kotlin DSL Kotlin Extensions sources from Java"
+    }
+
+    /**
+     * Defaults to `false`, set to `true` to use the embedded Gradle Kotlin DSL provider.
+     */
+    @Input
+    val isUseEmbeddedKotlinDslProvider = project.objects.property<Boolean>().apply {
+        set(false)
+    }
+
+    @Input
+    val nameComponents = project.objects.listProperty<String>().apply {
+        set(listOf(name))
+    }
+
+    @InputFiles
+    @Classpath
+    val classes = project.files()
+
+    @InputFiles
+    @Classpath
+    val classpath = project.files()
+
+    @Input
+    val includes = project.objects.listProperty<String>()
+
+    @Input
+    val excludes = project.objects.listProperty<String>()
+
+    @InputFiles
+    @PathSensitive(PathSensitivity.NONE)
+    val parameterNamesIndices = project.files()
+
+    @OutputDirectory
+    @PathSensitive(PathSensitivity.RELATIVE)
+    val outputDirectory = project.layout.directoryProperty()
+
+    @TaskAction
+    @Suppress("unused")
+    internal
+    fun generateKotlinDslApiExtensions() =
+        if (isUseEmbeddedKotlinDslProvider.get()) runEmbedded()
+        else runIsolated()
```

Comment:
What's the advantage of doing this? Doesn't the class-loader have to load the method `generateKotlinDslApiExtensionsSourceTo` either way simply because `GenerateKotlinDslApiExtensions` is loaded?

Reply:
`runEmbedded()` is useful only for development and testing purpose, and is not the default.
`runIsolated()` allows to use a version of the extension generation code different from the version of the embedded kotlin-dsl running the build where this plugin is applied.